File: Dsa-day55.java
import java.util.*;

class Solution {
    public int minimumTeachings(int n, int[][] languages, int[][] friendships) {
        int m = languages.length;
        List<Set<Integer>> langSets = new ArrayList<>();
        for (int[] langs : languages) {
            Set<Integer> set = new HashSet<>();
            for (int l : langs) set.add(l);
            langSets.add(set);
        }
        Set<Integer> needTeach = new HashSet<>();
        for (int[] f : friendships) {
            int u = f[0]-1, v = f[1]-1;
            boolean canCommunicate = false;
            for (int l : langSets.get(u)) {
                if (langSets.get(v).contains(l)) { canCommunicate = true; break; }
            }
            if (!canCommunicate) {
                needTeach.add(u);
                needTeach.add(v);
            }
        }
        if (needTeach.isEmpty()) return 0;
        int ans = Integer.MAX_VALUE;
        for (int lang = 1; lang <= n; lang++) {
            int count = 0;
            for (int user : needTeach) {
                if (!langSets.get(user).contains(lang)) count++;
            }
            ans = Math.min(ans, count);
        }
        return ans;
    }
}

File: Dsa-day55.cpp
class Solution {
public:
    int minimumTeachings(int n, vector<vector<int>>& languages, vector<vector<int>>& friendships) {
        int m = languages.size();
        vector<unordered_set<int>> langSets(m);
        for (int i = 0; i < m; i++)
            langSets[i] = unordered_set<int>(languages[i].begin(), languages[i].end());
        unordered_set<int> needTeach;
        for (auto &f : friendships) {
            int u = f[0]-1, v = f[1]-1;
            bool canCommunicate = false;
            for (int l : langSets[u]) {
                if (langSets[v].count(l)) { canCommunicate = true; break; }
            }
            if (!canCommunicate) {
                needTeach.insert(u);
                needTeach.insert(v);
            }
        }
        if (needTeach.empty()) return 0;
        int ans = INT_MAX;
        for (int lang = 1; lang <= n; lang++) {
            int count = 0;
            for (int user : needTeach) {
                if (!langSets[user].count(lang)) count++;
            }
            ans = min(ans, count);
        }
        return ans;
    }
};
